group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -903,6 +906,45 @@ public boolean cleanupExpiredOffsets(String groupId, List<Record> records) {
         return allOffsetsExpired.get() && !openTransactionsByGroup.containsKey(groupId);
     }
 
+    /**
+     * Remove offsets of the partitions that have been deleted.
+     *
+     * @param topicPartitions   The partitions that have been deleted.
+     * @return The list of tombstones (offset commit) to append.
+     */
+    public List<Record> onPartitionsDeleted(
```

Comment:
If we get a request to commit an offset that was just deleted (say the request comes in after the topic is deleted) would we just add it back to the offsets map?

Reply:
Yes, that's right. I think that there are two cases to consider here.

(1) The topic could have been deleted and recreated. In this case, this scenario is actually legit.

(2) The topic could have been deleted but there is a lingering offset commit request coming afterwards. We do have a check in KafkaApis which verifies if a partition actually exist before delegating the process of the request to the group coordinator. Note that this is not perfect because it is possible that a partition still exists when the verification is executed but does not exist anymore when the offset commit is processed by the group coordinator. This is also true in the old group coordinator.

As you already know, we will add support for topic ids soon. I think that this is the ultimate solution to those race conditions. When we have them, I think that we could think about dropping this deletion path and rely on the expiration logic only as we could expire offsets with unknown topic ids.

Reply:
Thanks for the thorough response. Makes sense to me. 👍